text/3529-cargo-path-bases.md

```diff
@@ -0,0 +1,426 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you often use multiple path dependencies that have a common parent directory,
+or if you want to avoid putting long paths in your `Cargo.toml`, you can
+define path _base directories_ in your
+[configuration](https://doc.rust-lang.org/cargo/reference/config.html).
+Your path dependencies can then be specified relative to those base
+directories.
+
+For example, say you have a number of projects checked out in
+`/home/user/dev/rust/libraries/`. Rather than use that path in your
+`Cargo.toml` files, you can define it as a "base" path in
+`~/.cargo/config.toml`:
+
+```toml
+[path-bases]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+Now, you can specify a path dependency on a library `foo` in that
+directory in your `Cargo.toml` using
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Like with other path dependencies, keep in mind that both the base _and_
+the path must exist on any other host where you want to use the same
+`Cargo.toml` to build your project.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Specifying Dependencies
+
+### Path Bases
+
+A `path` dependency may optionally specify a base by setting the `base` key to
+the name of a path base from the `[path-bases]` table in either the
+[configuration](https://doc.rust-lang.org/cargo/reference/config.html#path-bases)
+or one of the [built-in path bases](#built-in-path-bases). The value of that
+path base is prepended to the `path` value to produce the actual location where
+Cargo will look for the dependency.
+
+For example, if the Cargo.toml contains:
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Given a `[path-bases]` table in the configuration that contains:
+
+```toml
+[path-bases]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+This will produce a `path` dependency `foo` located at
+`/home/user/dev/rust/libraries/foo`.
+
+Path bases can be either absolute or relative. Relative path bases are relative
+to the parent directory of the configuration file that declared that path base.
+
+The name of a path base must use only [alphanumeric](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphanumeric)
+characters or `-` or `_`, must start with an [alphabetic](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphabetic)
+character, and must not be empty.
+
+If the name of path base used in a dependency is neither in the configuration
+nor one of the built-in path base, then Cargo will raise an error.
+
+#### Built-in path base
+
+Cargo provides implicit path bases that can be used without the need to specify
+them in a `[path-bases]` table.
+
+* `workspace` - If a project is [a workspace or workspace member](https://doc.rust-lang.org/cargo/reference/workspaces.html)
+then this path base is defined as the parent directory of the root Cargo.toml of
+the workspace.

```

Comment:
What about a `package` base that is the default value for `base`?

While this won't necessarily be a value users specify explicitly, I think it can help in modeling this to talk about how this feature works generally and not just in this specific case.

I am fine with this being an Unresolved Question that is left for stabilization.

Reply:
Hmm, if including config, maybe `current-dir` would be a better name

Reply:
Added these to the "unresolved questions" section, along with a few more suggestions.